src/smc-webapp/editor_chat.cjsx

```diff
@@ -699,11 +702,24 @@ ChatRoom = (name) -> rclass
             e.preventDefault()
             @clear_input()
         else if e.keyCode==13 and e.shiftKey # 13: enter key
-            @send_chat(e)
+            @setState(realtime_typing_message_id : '')
+            if not @state['realtime_typing_message_id']
+                @send_chat(e)
+            else
+                @clear_input()
         else if e.keyCode==38 and @refs.input.getValue() == ''
             # Up arrow on an empty input
             @props.actions.set_to_last_input()
 
+    keyup : (e) ->
+        if @state['realtime_typing']
+            if @state['realtime_typing_message_id']
+                @props.actions.send_edit(@props.messages.get(@state['realtime_typing_message_id']), @refs.input.getValue())
+            else
+                mesg = @refs.input.getValue()
+                if mesg.length? and mesg.trim().length >= 1
+                    @setState(realtime_typing_message_id:String(@props.actions.send_chat(mesg)))
```

Comment:
Using `send_chat` and `send_edit` is not the right way to go about doing this. The fact that this seems like a reasonable way to go is entirely my fault though. `send_chat` and `send_edit` were made for complete chats and complete edits. 

1. This will save every single key stroke into the history of the message. 
2. This will lock in your timestamp to **when you start typing** instead of when you send.

The right way to do this is to add a metadata item to the file that stores what each person currently has in their input box, whether they're actively typing (for some measurement of active), and if they want their message to be shown to everyone. This requires adding to the way the messages are scanned and the way changes are parsed in actions.